subprojects/scala/src/main/java/org/gradle/api/plugins/scala/ScalaPlugin.java

```diff
@@ -43,14 +49,33 @@ public void apply(Project project) {
         project.getPluginManager().apply(ScalaBasePlugin.class);
         project.getPluginManager().apply(JavaPlugin.class);
 
-        configureScaladoc(project);
+        final SourceSet main = project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets().getByName("main");
+
+        configureScaladoc(project, main);
+
+        final Configuration incrementalAnalysisElements = project.getConfigurations().getByName("incrementalScalaAnalysisElements");
+        String compileTaskName = main.getCompileTaskName("scala");
+        final TaskProvider<AbstractScalaCompile> compileScala = project.getTasks().withType(AbstractScalaCompile.class).named(compileTaskName);
+        final Provider<RegularFile> compileScalaMapping = project.getLayout().getBuildDirectory().file("tmp/scala/compilerAnalysis/" + compileTaskName + ".mapping");
+        compileScala.configure(new Action<AbstractScalaCompile>() {
+            @Override
+            public void execute(AbstractScalaCompile task) {
+                task.getAnalysisMappingFile().set(compileScalaMapping);
+            }
+        });
+        incrementalAnalysisElements.getOutgoing().artifact(
+            compileScalaMapping, new Action<ConfigurablePublishArtifact>() {
```

Comment:
Could the artifact be a mapping of `task.getAnalysisMappingFile()`?  Then you wouldn't have the issue of someone changing the value of `analysisMappingFile` on the task after you set it here.

Reply:
artifact(Object, Action) eagerly turns the first parameter into a file path, so it would force the task to be realized all the time.

Reply:
Hmm, I might be able to use the LazyArtifact thing...let me see.